File: src/za/co/adhd_developers/java/questions/Question17.java
package za.co.adhd_developers.java.questions;

import za.co.adhd_developers.java.tools.Question;

import java.util.Hashtable;

/**
 * Created by Grant on 2017/07/31.
 */
public class Question17 implements Question {

    Hashtable<Integer, String> staticStrings = new Hashtable<>();

    StringBuffer stringBuffer = new StringBuffer();

    public Question17() {
        this.staticStrings.put(1, "one");
        this.staticStrings.put(2, "two");
        this.staticStrings.put(3, "three");
        this.staticStrings.put(4, "four");
        this.staticStrings.put(5, "five");
        this.staticStrings.put(6, "six");
        this.staticStrings.put(7, "seven");
        this.staticStrings.put(8, "eight");
        this.staticStrings.put(9, "nine");
        this.staticStrings.put(10, "ten");
        this.staticStrings.put(11, "eleven");
        this.staticStrings.put(12, "twelve");
        this.staticStrings.put(13, "thirteen");
        this.staticStrings.put(14, "fourteen");
        this.staticStrings.put(15, "fifteen");
        this.staticStrings.put(16, "sixteen");
        this.staticStrings.put(17, "seventeen");
        this.staticStrings.put(18, "eighteen");
        this.staticStrings.put(19, "nineteen");
        this.staticStrings.put(20, "twenty");
        this.staticStrings.put(30, "thirty");
        this.staticStrings.put(40, "forty");
        this.staticStrings.put(50, "fifty");
        this.staticStrings.put(60, "sixty");
        this.staticStrings.put(70, "seventy");
        this.staticStrings.put(80, "eighty");
        this.staticStrings.put(90, "ninety");
        this.staticStrings.put(100, "one hundred");
        this.staticStrings.put(200, "two hundred");
        this.staticStrings.put(300, "three hundred");
        this.staticStrings.put(400, "four hundred");
        this.staticStrings.put(500, "five hundred");
        this.staticStrings.put(600, "six hundred");
        this.staticStrings.put(700, "seven hundred");
        this.staticStrings.put(800, "eight hundred");
        this.staticStrings.put(900, "nine hundred");
        this.staticStrings.put(1000, "one thousand");
    }

    @Override
    public void doWork() {
        for (int i = 1; i <= 1000; i++) {
            String number = String.valueOf(i);

            String finalString = "";

            if (this.staticStrings.containsKey(i)) {
                finalString = this.staticStrings.get(i);
            } else {
                int value = i;
                String strHundred = "";
                String strTens = "";
                String strValue = "";

                int hundred = value - (value % 100);
                value = value - hundred;

                if (this.staticStrings.containsKey(hundred)) {
                    strHundred = this.staticStrings.get(hundred);
                }

                if (this.staticStrings.containsKey(value)) {
                    strTens = this.staticStrings.get(value);
                } else {
                    int tens = value - (value % 10);
                    value = value - tens;

                    if (this.staticStrings.containsKey(tens)) {
                        strTens = this.staticStrings.get(tens);
                    }

                    if (this.staticStrings.containsKey(value)) {
                        strValue = this.staticStrings.get(value);
                    }
                }

                finalString = strHundred + (strHundred.length() > 0 ? " and" : "") + " " + strTens + " " + strValue;
            }

            finalString = finalString.trim().replace(" ", "");

            this.stringBuffer.append(finalString);
        }
    }

    @Override
    public void printAnswer() {
        System.out.println("=================================");
        System.out.println("Question 17");
        System.out.println("Number of letters of the numbers 1 - 1000 written out, is: " + this.stringBuffer.toString().length());
        System.out.println("=================================");
    }
}

File: src/za/co/adhd_developers/kotlin/questions/Question17.kt
package za.co.adhd_developers.kotlin.questions

import za.co.adhd_developers.kotlin.Question
import java.util.*

class Question17() : Question {

    var staticStrings = Hashtable<Int, String>()
    var stringBuilder = StringBuilder()

    init {
        this.staticStrings[1] = "one"
        this.staticStrings[2] = "two"
        this.staticStrings[3] = "three"
        this.staticStrings[4] = "four"
        this.staticStrings[5] = "five"
        this.staticStrings[6] = "six"
        this.staticStrings[7] = "seven"
        this.staticStrings[8] = "eight"
        this.staticStrings[9] = "nine"
        this.staticStrings[10] = "ten"
        this.staticStrings[11] = "eleven"
        this.staticStrings[12] = "twelve"
        this.staticStrings[13] = "thirteen"
        this.staticStrings[14] = "fourteen"
        this.staticStrings[15] = "fifteen"
        this.staticStrings[16] = "sixteen"
        this.staticStrings[17] = "seventeen"
        this.staticStrings[18] = "eighteen"
        this.staticStrings[19] = "nineteen"
        this.staticStrings[20] = "twenty"
        this.staticStrings[30] = "thirty"
        this.staticStrings[40] = "forty"
        this.staticStrings[50] = "fifty"
        this.staticStrings[60] = "sixty"
        this.staticStrings[70] = "seventy"
        this.staticStrings[80] = "eighty"
        this.staticStrings[90] = "ninety"
        this.staticStrings[100] = "onehundred"
        this.staticStrings[200] = "twohundred"
        this.staticStrings[300] = "threehundred"
        this.staticStrings[400] = "fourhundred"
        this.staticStrings[500] = "fivehundred"
        this.staticStrings[600] = "sixhundred"
        this.staticStrings[700] = "sevenhundred"
        this.staticStrings[800] = "eighthundred"
        this.staticStrings[900] = "ninehundred"
        this.staticStrings[1000] = "onethousand"
    }

    override fun doWork() {
        for (i in 1..1000) {
            var finalString: String

            if (this.staticStrings.containsKey(i)) {
                finalString = this.staticStrings[i]!!
            } else {
                var value = i
                var strHundred = ""
                var strTens = ""
                var strValue = ""

                val hundred = value - value % 100
                value -= hundred

                if (this.staticStrings.containsKey(hundred)) {
                    strHundred = this.staticStrings[hundred]!!
                }

                if (this.staticStrings.containsKey(value)) {
                    strTens = this.staticStrings[value]!!
                } else {
                    val tens = value - value % 10
                    value -= tens

                    if (this.staticStrings.containsKey(tens)) {
                        strTens = this.staticStrings[tens]!!
                    }

                    if (this.staticStrings.containsKey(value)) {
                        strValue = this.staticStrings[value]!!
                    }
                }

                finalString = strHundred + (if (strHundred.isNotEmpty()) "and" else "") + "" + strTens + "" + strValue
            }

            this.stringBuilder.append(finalString)
        }
    }

    override fun printAnswer() {
        println("=================================")
        println("Question 17")
        println("Number of letters of the numbers 1 - 1000 written out, is: " + this.stringBuilder.toString().length)
        println("=================================")
    }
}